### PR TITLE
Add sslsnoop and ssllatency tools

### PR DESCRIPTION
**Pull request by xutao323**
Add sslsnoop and ssllatency tools for SSL/TLS handshake performance analysis.

Use case and background is described here: https://xutao323.github.io/2022/01/06/tls-acceleration-with-new-CPU.html

### CODE REVIEW

**Comment by brendangregg** [comment]
Thanks; I think these are useful tools, but I don't think they'll work for everyone without edits to library paths. They just error for me as the paths can't be found. These would be the only tools in bpftrace that failed to run for most users, and would create a lot of tickets for people asking why. For that reason I wouldn't include them as is, as it will just give a lot of people a bad experience.

I'm not sure what to do, but I've been thinking along a few of different lines:

A) Include them in an examples directory instead and not the tools directory, and make it clear that this isn't expected to work without edits. bcc has an examples directory. I'm not sure this will be enough, people will still try to run them and complain.

B) Rewrite these in bcc libbpf-tools, where you can include a lot of other logic to search for the correct library paths and have the tool make a best effort to automatically do ssl tracing for you. 

C) Write this as a bash script that does the library and path checks, and finishes by invoking an embedded bpftrace program with the correct paths (or throws a better error message to say it can't find your ssl libraries). Programs that build programs are hard to debug and maintain, so my approach is to separate them as much as possible so it looks like this:

```
#!/bin/bash
...
bash logic
...
bpftrace -e '
...
bpftrace logic
...
'
```

So the tool source is in two halves, the top being bash and the bottom being bpftrace. A did this model years ago and maintaining it worked ok, for example: https://github.com/opendtrace/toolkit/blob/master/execsnoop

**Comment by xutao323** [comment]
Thanks for the reiview and suggestions. I'll add patch checks for the new tools.

At first I found sslsniff in bcc-tools will trace system ssl library (libssl.so) and it can resolve the library path automatically, so I also added SSL_do_handshake() trace to show crypto overhead in sslsniff [PR 3799](https://github.com/iovisor/bcc/pull/3799). As bpftrace is simpler to write and modify, I added those extra crypto function traces in bpftrace here as sslsnoop.bt and ssllatency.bt.

I'll try the "C" approach to combine library patch check and bpftrace script.

**Comment by fbs** [comment]
> Thanks; I think these are useful tools, but I don't think they'll work for everyone without edits to library paths. They just error for me as the paths can't be found. These would be the only tools in bpftrace that failed to run for most users, and would create a lot of tickets for people asking why. For that reason I wouldn't include them as is, as it will just give a lot of people a bad experience.

We don't have to specify the full path anymore so just doing 'u:libssl.so' should be enough :). We just haven't converted tools yet: #2075  it also has some caveats still: #2065

**Comment by xutao323** [comment]
> > Thanks; I think these are useful tools, but I don't think they'll work for everyone without edits to library paths. They just error for me as the paths can't be found. These would be the only tools in bpftrace that failed to run for most users, and would create a lot of tickets for people asking why. For that reason I wouldn't include them as is, as it will just give a lot of people a bad experience.
> 
> We don't have to specify the full path anymore so just doing 'u:libssl.so' should be enough :). We just haven't converted tools yet: #2075 it also has some caveats still: #2065

Sounds great. Then system library path like libssl.so and libcrypto.so can be automatically resolved. And existing uprobe tracing tools with same problem can be converted too like gethostlatency and threadsnoop.

One question about non-default library or binary, if they don't exist yet on some system (some already have), like libcrypto_mb.so or qatengine.so in /lib64/engines-1.1/, can we just skip those uprobes and only trace remaining common funcitons in libssl.so or libcrypto.so ? Otherwise, I'll need to check path or remove non-default uprobes in the proposed new tools which is the main analysis target with TLS performance optimization.

**Comment by fbs** [comment]
The docker container we build might not have it. Can you try adding `openssl` to the packages to install? You can add it to `docker/Dockerfile.focal` and `docker/Dockerfile.ubuntu-glibc`

**Comment by xutao323** [comment]
Embedded build test failed even after installed `openssl-dev` on `alpine`. I tried the docker build locally, and found it's a common `ldconfig` problem of `alpine linux` (ubuntu is fine). So lib name can't be resolved properly by `ldconfig` in `alpine`, it's suggested to `apk add libc6-compat or gcompat`. Shall we fix it here or open a new issue for it?

Test steps:
1. Build the docker `alpine` image according to the [build log](https://github.com/iovisor/bpftrace/runs/5123757531?check_suite_focus=true#step:3:2):
```
# docker build -t bpftrace-embedded-alpine -f docker/Dockerfile.alpine --build-arg bcc_ref= --build-arg BASE=alpine --build-arg LLVM_VERSION=9 --build-arg ALPINE_VERSION=3.11 docker/
```
2. Manually run test and it failed with no libs found: 
```
# docker run --privileged -v $(pwd):$(pwd) -w $(pwd) --entrypoint /bin/bash bpftrace-embedded-alpine -c "/root/bpftrace/build-embedded/src/bpftrace /root/bpftrace/build-embedded/tools/ssllatency.bt"
/root/bpftrace/build-embedded/tools/ssllatency.bt:17-19: ERROR: uprobe target file 'libssl' does not exist or is not executable
...
/root/bpftrace/build-embedded/tools/ssllatency.bt:41-44: ERROR: uprobe target file 'libcrypto' does not exist or is not executable
...
```
3. Use absolute path is OK:
```
# docker run --privileged -v $(pwd):$(pwd) -w $(pwd) --entrypoint /bin/bash bpftrace-embedded-alpine -c "/root/bpftrace/build-embedded/src/bpftrace -e 'uprobe:/lib/libssl.so.1.1:SSL_do_handshake { printf(\"enter handshake\n\"); }'"
Attaching 1 probe...
```
4. Even the glibc example in [manual](https://github.com/iovisor/bpftrace/blob/master/man/adoc/bpftrace.adoc#probes-uprobe) failed due to same problem:
```
# docker run --privileged -v $(pwd):$(pwd) -w $(pwd) --entrypoint /bin/bash bpftrace-embedded-alpine -c "/root/bpftrace/build-embedded/src/bpftrace -e 'uprobe:libc:malloc { printf(\"Allocated %d bytes\n\", arg0); }'"
stdin:1:1-19: ERROR: uprobe target file 'libc' does not exist or is not executable
uprobe:libc:malloc { printf("Allocated %d bytes\n", arg0); }
~~~~~~~~~~~~~~~~~~
# docker run --privileged -v $(pwd):$(pwd) -w $(pwd) --entrypoint /bin/bash bpftrace-embedded-alpine -c "/root/bpftrace/build-embedded/src/bpftrace -e 'uprobe:/lib/libc.so.6:malloc { printf(\"Allocated %d bytes\n\", arg0); }'"
Attaching 1 probe...
Allocated 4 bytes
Allocated 8 bytes
Allocated 8 bytes
Allocated 48 bytes
```
Next:
Install the suggested `libc6-compat or gcompat` , so that ldconfig can work properly in `alpine`.

**Comment by fbs** [comment]
for the alpine one we should probably just skip it, alpine is always a pain.  I think we already have the code to skip it but I'd have to check

**Comment by xutao323** [comment]
> for the alpine one we should probably just skip it, alpine is always a pain. I think we already have the code to skip it but I'd have to check

Any update ? Thanks in advance.

**Comment by xutao323** [comment]
#2156 is merged, so next step is to rerun the build tests ? if everything is ok, this new ssl/tls tool can be used like in bcc.

**Comment by viktormalik** [comment]
> #2156 is merged, so next step is to rerun the build tests ? if everything is ok, this new ssl/tls tool can be used like in bcc.

You'll need to disable tests for the new tools on Alpine (similarly to #2181).

**Comment by viktormalik** [comment]
Could you please rebase onto master and remove all the unnecessary merge commits? Thanks!

**Comment by xutao323** [comment]
Rebased onto most recent master and cleaned up the commits, please help to review, thanks.

File: tools/ssllatency.bt
Comment (viktormalik) [inline tools/ssllatency.bt:43]: Just wondering - what does this comment mean? Do you need some info from DWARF?

**Review by viktormalik — APPROVED** [review]

File: tools/ssllatency.bt
Comment (xutao323) [reply tools/ssllatency.bt:43]: The uprobes with "rsa_ossl" prefix are static functions in libcrypto.so that needs DWARF debug info. So *openssl-libs-debuginfo* should be installed for those uprobes in CentOS/Fedora , and *libssl-dev-dbgsym* in Ubuntu.

The rest uprobes in the same *action block* are fine without DWARF. bpftrace allows wildcard to match multiple attach points, and I tested run with no error if no DWARF package installed. I think it shall be fine as far as there's at least one matching uprobe found.

**Comment by xutao323** [comment]
any update? waiting for approval or any additional commennts need to be addressed? thanks.

**Comment by viktormalik** [comment]
@brendangregg the tools should now work for most of the systems, do you agree with merging this?

**Comment by fbs** [comment]
Sorry for the radio silence here. I think they're useful so lets just get them in. In the end they're only example tools we ship so we shouldn't make it too difficult, worst case we can delete them again.

Can you rabse to fix the few conflicts?

File: CHANGELOG.md
Comment (fbs) [inline CHANGELOG.md:71]: argh sorry for the extra work but can you move this entry up? We did the 0.15 release and this needs to go into `unreleased`

Comment (xutao323) [reply CHANGELOG.md]: done

**Pull request merged**
